File: practical_works/perceptron/Python/iris/iris_modular.py


# Neural networks pedagogical materials
# The following code is free to use and modify to any extent 
# (with no responsibility of the original author)

# Reference to the author is a courtesy
# Author : David Cornu => david.cornu@utinam.cnrs.fr



import numpy as np
import sys
sys.path.insert(0,'..')
from ext_module import *


nb_epochs = 5000
learn_rate = 0.005


######################### ##########################
#          Loading data and pre process
######################### ##########################

raw_data = np.loadtxt("../../../data/iris.data")


nb_dat = np.shape(raw_data)[0]
in_dim = np.shape(raw_data)[1] - 1
out_dim = 3

input = np.append(raw_data[:,:in_dim], -1.0*np.ones((nb_dat,1)), axis=1)
output = np.zeros(out_dim)

targ = np.zeros((nb_dat,out_dim))
for i in range(0,nb_dat):
	targ[i,int(raw_data[i,in_dim])] = 1.0



input[:,:-1] -= np.mean(input[:,:-1], axis = 0)
input[:,:-1] /= np.max(np.abs(input[:,:-1]), axis = 0)

index = np.arange(0,nb_dat)

######################### ##########################
#          Initialize network weights
######################### ##########################

weights = np.random.rand(in_dim+1,out_dim)*(0.02)-0.01


######################### ##########################
#                Main training loop
######################### ##########################
######################### ##########################

for t in range(0,nb_epochs):

	if (((t%200) == 0)):
		print("\nIteration :", t)
		accu = 0.0
		
		for n in range(0,nb_dat):
			forward(input[n,:], output, weights)

			if(np.all(output == targ[n])):
				accu += 1
				
		print ("  Accu %8.2f%%"%(accu/nb_dat*100.0))

	np.random.shuffle(index)
	input = input[index,:]
	targ = targ[index,:]

	######################### ##########################
	#             Training on all data once
	######################### ##########################
	for i in range(0,nb_dat):
		#Forward phase
		forward(input[i,:], output, weights)
		
		#Back-propagation phase
		backprop(input[i,:], output, targ[i,:], weights, learn_rate)


######################### ##########################
######################### ##########################

File: practical_works/perceptron/Fortran/iris/iris_modular.f90



! Neural networks pedagogical materials
! The following code is free to use and modify to any extent 
! (with no responsibility of the original author)

! Reference to the author is a courtesy
! Author : David Cornu => david.cornu@observatoiredeparis.psl.eu


program logic_or
	
	use ext_module

	implicit none

	integer, parameter :: nb_dat = 150, in_dim = 4, out_dim = 3, nb_epochs = 5000
	real, dimension(nb_dat,in_dim+1) :: input
	real, dimension(nb_dat,out_dim) :: targ
	real, dimension(out_dim) :: output
	real :: h, learn_rate = 0.005
	real, dimension(in_dim+1,out_dim) :: weights
	integer :: i, t, ind, temp_class
	


	!######################## ##########################
	!          Loading data and pre process
	!######################## ##########################
	
	open(10, file="../../../data/iris.data")
	
	targ(:,:) = 0
	do i=1, nb_dat
		read(10,*) input(i,1:4), temp_class
		targ(i, temp_class+1) = 1.0
	end do
	input(:,5) = -1.0

	do i=1, 4
		input(:,i) = input(:,i)-(sum(input(:,i))/nb_dat)
		input(:,i) = input(:,i)/maxval(abs(input(:,i)))
	end do


	
	!######################## ##########################
	!          Initialize network weights
	!######################## ##########################

	call random_number(weights)
	weights(:,:) = weights(:,:)*(0.02)-0.01
	
	
	!######################## ##########################
	!                Main training loop
	!######################## ##########################
	!######################## ##########################
	
	do t = 1, nb_epochs
	
		if (mod(t,200) == 0 .OR. t == 1) then
			Write(*,*)
			write(*,*) "Iteration :", t
		
			call confmat(input(:,:), in_dim, targ(:,:), out_dim, nb_dat, weights)
		end if

		call shuffle(input, in_dim+1, targ, out_dim, nb_dat)

		!######################## ##########################
		!             Training on all data once
		!######################## ##########################
		do i=1, nb_dat
			!Forward phase
			call forward(input(i,:), in_dim, output, out_dim, weights)
			
			!Back-propagation phase
			call backprop(input(i,:), in_dim, output, targ(i,:), out_dim, weights, learn_rate)
		end do
	end do
	
	!######################## ##########################
	!######################## ##########################
	
end program logic_or














